File: power-adapters-data/src/main/java/com/nextfaze/poweradapters/data/FakeData.java
package com.nextfaze.poweradapters.data;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import com.nextfaze.poweradapters.internal.NotificationType;
import com.nextfaze.poweradapters.internal.NotifyingArrayList;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;

@SuppressWarnings("WeakerAccess")
@VisibleForTesting
public class FakeData<T> extends Data<T> {

    @NonNull
    private final NotifyingArrayList<T> mData = new NotifyingArrayList<>(mDataObservable);

    private int mAvailable = UNKNOWN;

    private boolean mLoading;

    @Override
    public int size() {
        return mData.size();
    }

    @NonNull
    @Override
    public T get(int position, int flags) {
        //noinspection ConstantConditions
        return mData.get(position);
    }

    public boolean add(@NonNull T t) {
        return mData.add(t);
    }

    public boolean addAll(@NonNull Collection<? extends T> collection) {
        return mData.addAll(collection);
    }

    public boolean addAll(int index, @NonNull Collection<? extends T> collection) {
        return mData.addAll(index, collection);
    }

    public void clear() {
        mData.clear();
    }

    public boolean remove(@NonNull T t) {
        return mData.remove(t);
    }

    public T set(int index, T object) {
        return mData.set(index, object);
    }

    @Override
    public void invalidate() {
    }

    @Override
    public void refresh() {
    }

    @Override
    public void reload() {
    }

    @Override
    public int available() {
        return mAvailable;
    }

    @NonNull
    @Override
    public List<T> asList() {
        return mData;
    }

    public void setAvailable(int available) {
        if (available != mAvailable) {
            mAvailable = available;
            notifyAvailableChanged();
        }
    }

    @Override
    public boolean isLoading() {
        return mLoading;
    }

    public void setLoading(boolean loading) {
        if (loading != mLoading) {
            mLoading = loading;
            notifyLoadingChanged();
        }
    }

    @SuppressWarnings("unchecked")
    public void insert(int index, @NonNull T... items) {
        mData.addAll(index, Arrays.asList(items));
    }

    @SuppressWarnings("unchecked")
    public void append(@NonNull T... items) {
        addAll(Arrays.asList(items));
    }

    @SuppressWarnings("unchecked")
    public void change(int index, @NonNull T... items) {
        mData.setAll(index, Arrays.asList(items));
    }

    public void remove(int index, int count) {
        mData.remove(index, count);
    }

    public void move(int fromPosition, int toPosition, int itemCount) {
        mData.move(fromPosition, toPosition, itemCount);
    }

    public void error(@NonNull Throwable e) {
        super.notifyError(e);
    }

    @SuppressWarnings("unused")
    @NonNull
    public NotificationType getNotificationType() {
        return mData.getNotificationType();
    }

    public void setNotificationType(@NonNull NotificationType notificationType) {
        mData.setNotificationType(notificationType);
    }
}

File: power-adapters-test/src/main/java/com/nextfaze/poweradapters/data/test/FakeData.kt
package com.nextfaze.poweradapters.data.test

import com.nextfaze.poweradapters.data.Data
import com.nextfaze.poweradapters.internal.NotificationType
import com.nextfaze.poweradapters.internal.NotifyingArrayList

class FakeData<T> : Data<T>() {

    private val data = NotifyingArrayList<T>(dataObservable)

    var available = Data.UNKNOWN
        set(value) {
            if (value != field) {
                field = value
                notifyAvailableChanged()
            }
        }

    var loading = false
        set(value) {
            if (value != field) {
                field = value
                notifyLoadingChanged()
            }
        }

    var notificationType: NotificationType
        get() = data.notificationType
        set(notificationType) {
            data.notificationType = notificationType
        }

    override fun size(): Int = data.size

    override fun get(position: Int, flags: Int): T = data[position]

    fun add(t: T) = data.add(t)

    fun addAll(collection: Collection<T>) = data.addAll(collection)

    fun addAll(index: Int, collection: Collection<T>) = data.addAll(index, collection)

    fun clear() = data.clear()

    fun remove(t: T) = data.remove(t)

    operator fun set(index: Int, o: T): T = data.set(index, o)

    override fun invalidate() {
    }

    override fun refresh() {
    }

    override fun reload() {
    }

    override fun available() = available

    override fun isLoading() = loading

    override fun asList() = data

    fun insert(index: Int, vararg items: T) = data.addAll(index, listOf(*items))

    fun append(vararg items: T) = addAll(listOf(*items))

    fun change(index: Int, vararg items: T) = data.setAll(index, listOf(*items))

    fun remove(index: Int, count: Int) = data.remove(index, count)

    fun move(fromPosition: Int, toPosition: Int, itemCount: Int) = data.move(fromPosition, toPosition, itemCount)

    fun error(e: Throwable) {
        super.notifyError(e)
    }
}
